src/charm_state.py

```diff
@@ -0,0 +1,199 @@
+    ) -> None:
+        """Construct.
+
+        Args:
+            synapse_config: The value of the synapse_config charm configuration.
+            datasource: Datasource information.
+            saml_config: SAML configuration.
+        """
+        self._synapse_config = synapse_config
+        self._datasource = datasource
+        self._saml_config = saml_config
+
+    @property
+    def server_name(self) -> typing.Optional[str]:
+        """Return server_name config.
+
+        Returns:
+            str: server_name config.
+        """
+        return self._synapse_config.server_name
+
+    @property
+    def report_stats(self) -> typing.Union[str, bool, None]:
+        """Return report_stats config.
+
+        Returns:
+            str: report_stats config as yes or no.
+        """
+        return self._synapse_config.report_stats
+
+    @property
+    def public_baseurl(self) -> typing.Optional[str]:
+        """Return public_baseurl config.
+
+        Returns:
+            str: public_baseurl config.
+        """
+        return self._synapse_config.public_baseurl
+
+    @property
+    def enable_mjolnir(self) -> bool:
+        """Return enable_mjolnir config.
+
+        Returns:
+            bool: enable_mjolnir config.
+        """
+        return self._synapse_config.enable_mjolnir
+
+    @property
+    def datasource(self) -> typing.Union[DatasourcePostgreSQL, None]:
+        """Return datasource.
+
+        Returns:
+            datasource or None.
+        """
+        return self._datasource
+
+    @property
+    def saml_config(self) -> typing.Union[SAMLConfiguration, None]:
+        """Return SAML configuration.
+
+        Returns:
+            SAMLConfiguration or None.
+        """
+        return self._saml_config
+
+    @classmethod
+    def from_charm(cls, charm: "SynapseCharm") -> "CharmState":
+        """Initialize a new instance of the CharmState class from the associated charm.
+
+        Args:
+            charm: The charm instance associated with this state.
+
+        Return:
+            The CharmState instance created by the provided charm.
+
+        Raises:
+            CharmConfigInvalidError: if the charm configuration is invalid.
+        """
+        synapse_config = {k: v for k, v in charm.config.items() if k in KNOWN_CHARM_CONFIG}
+        try:
+            valid_synapse_config = SynapseConfig(**synapse_config)  # type: ignore
+        except ValidationError as exc:
+            error_fields = set(
+                itertools.chain.from_iterable(error["loc"] for error in exc.errors())
+            )
+            error_field_str = " ".join(f"{f}" for f in error_fields)
+            raise CharmConfigInvalidError(f"invalid configuration: {error_field_str}") from exc
+        return cls(
+            synapse_config=valid_synapse_config,
+            datasource=charm.database.get_relation_as_datasource(),
+            saml_config=charm.saml.get_relation_as_saml_conf(),
```

Comment:
The Flask charm was used as reference regarding passing SynapseCharm instead of CharmBase
https://github.com/canonical/flask-k8s-operator/blob/e485689ae9774874f1e95b154ac79831ea352fd2/src/charm_state.py#L188
```
def from_charm(
        cls, charm: "FlaskCharm", secret_storage: SecretStorage, database_uris: dict[str, str]
    ) -> "CharmState":
```
I agree that DatasourcePostgreSQL and SAMLConfiguration could be passed as arguments.

Reply:
Related: #54

Reply:
Yeh we have a similar refactoring that will probably be needed in the flask charm